sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/CachedTokenProvider.java

```diff
@@ -0,0 +1,49 @@
+package com.mesosphere.sdk.dcos.auth;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Optional;
+
+/**
+ * CachedTokenProvider retrieves token from underlying provider and caches the value. It automatically triggers
+ * getToken() method on underlying provider when token is about to expire.
+ *
+ * // TODO(mh): This could be also running in a separate thread in background.
+ * // TODO(mh): Does this needs to be threadsafe?
+ */
+public class CachedTokenProvider implements TokenProvider {
+
+    private TokenProvider provider;
```

Comment:
mark `provider` and `triggerRefreshBeforeSeconds` (or `ttl`, see below) as `final`